content/blog/2023/2023-11-23-sorting-basics-kotlin.md

```diff
@@ -0,0 +1,108 @@
+
+    val sortedNumbers = numbersArray.sorted()
+    println("Sorted numbers: $sortedNumbers")
+
+    val descendingNumbers = numbersArray.sortedDescending()
+    println("Descending numbers: $descendingNumbers")
+}
+
+```
+
+In our code above, we initialize an array named `numbersArrays` with integers and demonstrate sorting operations. First, it employs the sorted() function to create a new array `sortedArrayNumbers` containing the elements of numbers in ascending order. The sorted array is then printed to the console using contentToString(). Subsequently, the sortedDescending() function is applied to obtain another array, descendingNumbers, with the elements sorted in descending order and this array is printed as well.
+
+
+## SortBy
+
+The sortBy function in Kotlin is used to sort a collection for example a list or an array based on a specified key or custom sorting criteria.
+
+ Here's an example using sortBy:
+
+ ```kotlin
+ data class Person(val name: String, val position: Int)
+
+fun main() {
+    val people = listOf(
+        Person("John", 1),
+        Person("Doe", 2),
+        Person("Mary",3)
+    )
+
+    // Sorting people by position in ascending order
+    val sortedPeople = people.sortedBy { it.position }
+    println("Sorted by position: $sortedPeople")
+
+    // Sorting people by position in descending order
+    val reverseSortedPeople = people.sortedByDescending { it.position }
+    println("Reverse sorted by position: $reverseSortedPeople")
+}
+ ```
+
+ In this example, a `Person` data class is defined with a `name` and `position`. The people list is then sorted using the `sortedBy` function specifying that the sorting should be based on the position property. The result is a new list sortedPeople, where the people are sorted in ascending order of position. Similarly, `sortedByDescending` is used to sort in descending order.
+
+ ## sortWith
+
+ The sortWith function in Kotlin allows us to provide a custom comparator to define how elements in a collection should be compared and sorted.
+
+ Here's an example using sortWith:
+ ```kotlin
+ data class Book(val title: String, val author: String, val publicationYear: Int)
+
+fun main() {
+    val books = listOf(
+        Book("The Great Gatsby", "F. Scott Fitzgerald", 1925),
+        Book("To Kill a Mockingbird", "Harper Lee", 1960),
+        Book("1984", "George Orwell", 1949),
+        Book("The Catcher in the Rye", "J.D. Salinger", 1951)
+    )
+
+    // Sorting books by publication year in ascending order using sortWith and a custom comparator
+    val sortedBooks = books.sortedWith(compareBy { it.publicationYear })
+    println("Books sorted by publication year: $sortedBooks")
+
+    // Sorting books by publication year in descending order using sortWith and a custom comparator
+    val reverseSortedBooks = books.sortedWith(compareByDescending { it.publicationYear })
+    println("Books reverse sorted by publication year: $reverseSortedBooks")
```

Comment:
Same as above: another example with a bit more complicated sorting, like "first sort by name, then sort by year" would be great.